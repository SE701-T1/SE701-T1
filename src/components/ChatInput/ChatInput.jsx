import React, { useState } from 'react';
import { IconButton, InputBase } from '@mui/material';
import SendIcon from '@mui/icons-material/Send';
import EmojiEmotionsIcon from '@mui/icons-material/EmojiEmotions';
import { styled } from '@mui/material/styles';
import Picker from 'emoji-picker-react';
import PropTypes from 'prop-types';
import styles from './ChatInput.module.css';

const CustomInput = styled(InputBase)(() => ({
  '& .MuiInputBase-input': {
    height: 'auto',
    fontSize: 16,
    color: '#7A7A7A',
    padding: 0,
  },
}));

/**
 * ChatInput is the component that handles the input of the user in the chat.
 * This component includes the emoji selector and submit button.
 */
function ChatInput({ onSend, disable }) {
  const [inputText, setInputText] = useState('');
  const [showEmojiPicker, setShowEmojiPicker] = useState(false);

  const onEmojiClick = (event, emojiObject) => {
    setInputText(inputText + emojiObject.emoji);
  };

  const onSubmit = () => {
    onSend(inputText);
    setInputText('');
  };

  const onKeyPress = (event) => {
    if (event.key === 'Enter') {
      onSubmit();
    }
  };

  return (
    <div className={styles.container}>
      <CustomInput
        disabled={disable}
        className={styles.input}
        inputRef={(input) => {
          if (input != null) {
            input.focus();
          }
        }}
        placeholder="Aa"
        value={inputText}
        onChange={(e) => setInputText(e.target.value)}
        onKeyDown={onKeyPress}
      />
      <div className={styles.buttonContainer}>
        <IconButton
          disabled={disable}
          data-testid="EmojiButton"
          className={styles.button}
          aria-label="directions"
          onClick={() => setShowEmojiPicker(!showEmojiPicker)}>
          <EmojiEmotionsIcon />
        </IconButton>
        {showEmojiPicker && (
          <div data-testid="EmojiSelector" className={styles.emojiButton}>
            <Picker disableSearchBar disableSkinTonePicker onEmojiClick={onEmojiClick} />
          </div>
        )}
      </div>
      <IconButton
        disabled={disable}
        data-testid="SendButton"
        className={styles.button}
        aria-label="send"
        onClick={onSubmit}>
        <SendIcon />
      </IconButton>
    </div>
  );
}

ChatInput.propTypes = {
  onSend: PropTypes.func.isRequired,
};

export default ChatInput;
