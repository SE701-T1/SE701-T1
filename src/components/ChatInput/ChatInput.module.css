.container {
  display: flex;
  align-items: center;
  justify-content: center;
}

.buttonContainer {
  position: relative;
}

.button {
  margin-left: 10px;
}

.emojiButton {
  position: absolute;
  bottom: 50px;
  right: 10px;
}

.input {
  border-radius: 25px;
  background-color: #f0f0f0;
  padding: 5px 15px;
  flex: 1;
}
