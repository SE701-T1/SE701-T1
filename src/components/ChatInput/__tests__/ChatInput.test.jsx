import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ChatInput from '../ChatInput';

describe('ChatInput', () => {
  it('is correctly rendered to the screen', () => {
    render(<ChatInput onSend={() => null} />);

    expect(screen.getByPlaceholderText('Aa')).toBeInTheDocument();
    expect(screen.getByTestId('EmojiEmotionsIcon')).toBeInTheDocument();
  });

  it('Emoji selector correctly opens', () => {
    render(<ChatInput onSend={() => null} />);

    fireEvent.click(screen.getByTestId('EmojiButton'));

    expect(screen.getByTestId('EmojiSelector')).toBeInTheDocument();
  });

  it('Emoji selector correctly opens', () => {
    render(<ChatInput onSend={() => null} />);

    fireEvent.click(screen.getByTestId('EmojiButton'));

    expect(screen.getByTestId('EmojiSelector')).toBeInTheDocument();
  });

  it('Send button clears text', () => {
    render(<ChatInput onSend={() => null} />);

    fireEvent.change(screen.getByPlaceholderText('Aa'), {
      target: {
        value: 'test',
      },
    });
    fireEvent.click(screen.getByTestId('SendButton'));

    expect(screen.getByPlaceholderText('Aa')).toHaveValue('');
  });

  it('produces the correct initial snapshot', () => {
    const tree = renderer.create(<ChatInput onSend={() => null} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
