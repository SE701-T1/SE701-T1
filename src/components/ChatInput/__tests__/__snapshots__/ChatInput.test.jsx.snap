// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatInput produces the correct initial snapshot 1`] = `
<div
  className="container"
>
  <div
    className="MuiInputBase-root MuiInputBase-colorPrimary input css-1ki836k-MuiInputBase-root"
    onClick={[Function]}
  >
    <input
      className="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="Aa"
      type="text"
      value=""
    />
  </div>
  <div
    className="buttonContainer"
  >
    <button
      aria-label="directions"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium button css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      data-testid="EmojiButton"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="EmojiEmotionsIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM8.5 8c.83 0 1.5.67 1.5 1.5S9.33 11 8.5 11 7 10.33 7 9.5 7.67 8 8.5 8zM12 18c-2.28 0-4.22-1.66-5-4h10c-.78 2.34-2.72 4-5 4zm3.5-7c-.83 0-1.5-.67-1.5-1.5S14.67 8 15.5 8s1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
        />
      </svg>
    </button>
  </div>
  <button
    aria-label="send"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium button css-78trlr-MuiButtonBase-root-MuiIconButton-root"
    data-testid="SendButton"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="SendIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
      />
    </svg>
  </button>
</div>
`;
